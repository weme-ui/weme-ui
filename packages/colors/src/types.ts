export type ColorAppearance = 'light' | 'dark'

export type ColorType = 'accent' | 'neutral'

export type ColorScalesArray<T = string> = [T, T, T, T, T, T, T, T, T, T, T, T]

export type Colors<T = string> = Record<string, ColorScalesArray<T>>
