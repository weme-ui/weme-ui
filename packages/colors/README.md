<p align="center">
  <img align="center" src="https://raw.githubusercontent.com/moujinet/assets/main/weme-ui/png/weme-128-duotone.png" height="128" />
  <h1 align="center">
    Weme UI - Colors
  </h1>
</p>

[![npm version][npm-version-src]][npm-version-href]
[![License][license-src]][license-href]

## 特性

- 🎨 [Radix Colors](https://www.radix-ui.com/colors) 基于 Radix Colors 并提供便捷的自定义颜色

## 如何开始

```shell
pnpm i -D @weme-ui/colors
```

## 许可证

[MIT][license-href] License © 2025 [weme-ui][github-href]

[github-href]: https://github.com/weme-ui/weme-ui
[npm-version-src]: https://img.shields.io/npm/v/@weme-ui/colors?style=flat&colorA=18181b&colorB=18181b
[npm-version-href]: https://npmjs.com/package/@weme-ui/colors
[license-src]: https://img.shields.io/github/license/@weme-ui/weme-ui.svg?style=flat&colorA=18181b&colorB=18181b
[license-href]: https://github.com/weme-ui/weme-ui/blob/main/LICENSE
