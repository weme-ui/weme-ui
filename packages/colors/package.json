{
  "name": "@weme-ui/colors",
  "type": "module",
  "version": "0.0.1",
  "packageManager": "pnpm@10.13.1",
  "description": "Weme UI - Colors",
  "author": "Allen Luo <luoyi@mouji.net>",
  "license": "MIT",
  "homepage": "https://github.com/weme-ui/weme-ui#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/weme-ui/weme-ui.git",
    "directory": "packages/colors"
  },
  "bugs": {
    "url": "https://github.com/weme-ui/weme-ui/issues"
  },
  "keywords": [
    "weme-ui",
    "weme-ui-colors",
    "radix-colors",
    "radix-colors-generator"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js"
    }
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "typesVersions": {
    "*": {
      "*": [
        "./dist/*",
        "./dist/index.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsdown",
    "dev": "tsdown --watch",
    "dev:prepare": "pnpm run build",
    "prepublishOnly": "pnpm run build",
    "release": "bumpp --commit --push --no-tag && pnpm publish --no-git-checks --access public",
    "test": "vitest",
    "typecheck": "tsgo --noEmit"
  },
  "peerDependencies": {
    "@radix-ui/colors": "catalog:dev",
    "bezier-easing": "catalog:dev",
    "colorjs.io": "catalog:dev"
  },
  "devDependencies": {
    "@radix-ui/colors": "catalog:dev",
    "@types/node": "catalog:build",
    "bezier-easing": "catalog:dev",
    "bumpp": "catalog:build",
    "colorjs.io": "catalog:dev",
    "tsdown": "catalog:build",
    "typescript": "catalog:build",
    "vitest": "catalog:testing"
  }
}
