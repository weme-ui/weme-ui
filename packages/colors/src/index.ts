export * from './colors'
export * from './transform'
export * from './types'
export * from './utils'
