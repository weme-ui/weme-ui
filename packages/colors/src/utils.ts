import type { ColorAppearance, ColorScalesArray, ColorType } from './types'
import * as RadixColors from '@radix-ui/colors'
import Color from 'colorjs.io'

const SPECIAL_COLOR_NAMES = ['black', 'white'] as const
const NEUTRAL_COLOR_NAMES = ['gray', 'mauve', 'slate', 'sage', 'olive', 'sand'] as const
const ACCENT_COLOR_NAMES = ['tomato', 'red', 'ruby', 'crimson', 'pink', 'plum', 'purple', 'violet', 'iris', 'indigo', 'blue', 'cyan', 'teal', 'jade', 'green', 'grass', 'bronze', 'gold', 'brown', 'orange', 'amber', 'yellow', 'lime', 'mint', 'sky'] as const
const COLOR_NAMES = [...SPECIAL_COLOR_NAMES, ...NEUTRAL_COLOR_NAMES, ...ACCENT_COLOR_NAMES] as const

export type RadixColorNames = keyof typeof RadixColors

export type AllColorNames = typeof NEUTRAL_COLOR_NAMES[number] | typeof ACCENT_COLOR_NAMES[number]
export type AccentColorNames = typeof ACCENT_COLOR_NAMES[number]
export type NeutralColorNames = typeof NEUTRAL_COLOR_NAMES[number]

export function isRadixColorName(color: string) {
  return COLOR_NAMES.includes(color as any)
}

export function getColorNames(type?: ColorType) {
  return !type
    ? COLOR_NAMES
    : type === 'accent'
      ? ACCENT_COLOR_NAMES
      : NEUTRAL_COLOR_NAMES
}

export function getRadixColorName(name: string, appearance: ColorAppearance) {
  return name === 'black'
    ? 'blackP3A'
    : name === 'white'
      ? 'whiteP3A'
      : `${name}${appearance === 'light' ? 'P3' : 'DarkP3'}` as RadixColorNames
}

export function getRadixColor<
  S extends boolean = false,
  T extends S extends true ? string : Color = S extends true ? string : Color,
  RT = ColorScalesArray<T>,
>(
  appearance: ColorAppearance,
  name: string,
  stringify?: boolean,
): RT {
  const colorName = getRadixColorName(name, appearance)

  return Object.values(RadixColors[colorName]).map(
    color => stringify
      ? toColorString(color)
      : new Color(color).to('oklch'),
  ) as RT
}

export function toColorString(colorOrStr: Color | string): string {
  const color = typeof colorOrStr === 'string'
    ? new Color(colorOrStr)
    : colorOrStr

  return color
    .to('oklch')
    .toString({ precision: 4 })
}
