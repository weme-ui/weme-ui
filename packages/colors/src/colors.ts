import type Color from 'colorjs.io'
import type { ColorAppearance, Colors } from './types'
import { getColorNames, getRadixColor } from './utils'

const lightColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

const lightNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

export function getColors(appearance: ColorAppearance, isNeutral?: boolean) {
  return isNeutral
    ? appearance === 'light' ? lightNeutralColors : darkNeutralColors
    : appearance === 'light' ? lightColors : darkColors
}
