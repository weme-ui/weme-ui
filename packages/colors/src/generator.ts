import type { Colors, ColorScalesArray } from './types'
import BezierEasing from 'bezier-easing'
import Color from 'colorjs.io'
import { getColorNames, toColorString } from './utils'

export function generateColor(
  color: string,
  colors: Colors<Color>,
  background: string,
): ColorScalesArray {
  const sourceColor = new Color(color).to('oklch')
  const backgroundColor = new Color(background).to('oklch')

  const scales = getScaleColors(sourceColor, colors, backgroundColor)

  scales[8] = sourceColor
  scales[9] = getHoverColor(sourceColor, [scales])

  scales[10].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[10].coords[1],
  )

  scales[11].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[11].coords[1],
  )

  return scales.map(toColorString) as ColorScalesArray
}

function getScaleColors(
  source: Color,
  colors: Record<string, ColorScalesArray<Color>>,
  backgroundColor: Color,
): ColorScalesArray<Color> {
  const allColors: { name: string, color: Color, distance: number }[] = []

  Object.entries(colors).forEach(([name, scale]) => {
    for (const color of scale) {
      const distance = source.deltaEOK(color)
      allColors.push({ name, distance, color })
    }
  })

  allColors.sort((a, b) => a.distance - b.distance)

  const closestColors = allColors.filter(
    (color, i, arr) => i === arr.findIndex(value => value.name === color.name),
  )

  const grayColorNames = getColorNames('neutral') as readonly string[]
  const allAreGrays = closestColors.every(color =>
    grayColorNames.includes(color.name),
  )
  if (!allAreGrays && grayColorNames.includes(closestColors[0].name)) {
    while (grayColorNames.includes(closestColors[1].name)) {
      closestColors.splice(1, 1)
    }
  }

  const closestA = closestColors[0]
  const closestB = closestColors[1]

  const a = closestB.distance
  const b = closestA.distance
  const c = closestA.color.deltaEOK(closestB.color)

  const cosA = (b ** 2 + c ** 2 - a ** 2) / (2 * b * c)
  const radA = Math.acos(cosA)
  const sinA = Math.sin(radA)

  const cosB = (a ** 2 + c ** 2 - b ** 2) / (2 * a * c)
  const radB = Math.acos(cosB)
  const sinB = Math.sin(radB)

  const tanC1 = cosA / sinA
  const tanC2 = cosB / sinB
  const ratio = Math.max(0, tanC1 / tanC2) * 0.5

  const colorA = colors[closestA.name]
  const colorB = colors[closestB.name]

  const scale = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].map(i =>
    new Color(Color.mix(colorA[i], colorB[i], ratio)).to('oklch'),
  ) as ColorScalesArray<Color>

  const baseColor = scale.slice()
    .sort((a, b) => source.deltaEOK(a) - source.deltaEOK(b))[0]

  const ratioC = source.coords[1] / baseColor.coords[1]

  scale.forEach((color) => {
    color.coords[1] = Math.min(
      source.coords[1] * 1.5,
      color.coords[1] * ratioC,
    )
    color.coords[2] = source.coords[2]
  })

  // Light
  if (scale[0].coords[0] > 0.5) {
    const lightnessScale = scale.map(({ coords }) => coords[0])
    const backgroundL = Math.max(0, Math.min(1, backgroundColor.coords[0]))
    const newLightnessScale = transposeProgressionStart(
      backgroundL,
      [1, ...lightnessScale],
      [0, 2, 0, 2] as [number, number, number, number],
    )

    newLightnessScale.shift()

    newLightnessScale.forEach((lightness, i) => {
      scale[i].coords[0] = lightness
    })

    return scale
  }

  // Dark
  const ease: [number, number, number, number] = [1, 0, 1, 0]
  const referenceBackgroundColorL = scale[0].coords[0]
  const backgroundColorL = Math.max(0, Math.min(1, backgroundColor.coords[0]))

  const ratioL = backgroundColorL / referenceBackgroundColorL

  if (ratioL > 1) {
    const maxRatio = 1.5

    for (let i = 0; i < ease.length; i++) {
      const metaRatio = (ratioL - 1) * (maxRatio / (maxRatio - 1))
      ease[i] = ratioL > maxRatio ? 0 : Math.max(0, ease[i] * (1 - metaRatio))
    }
  }

  const lightnessScale = scale.map(({ coords }) => coords[0])
  const newLightnessScale = transposeProgressionStart(
    backgroundColorL,
    lightnessScale,
    ease,
  )

  newLightnessScale.forEach((lightness, i) => {
    scale[i].coords[0] = lightness
  })

  return scale
}

function getHoverColor(source: Color, scales: ColorScalesArray<Color>[]) {
  const [L, C, H] = source.coords
  const newL = L > 0.4 ? L - 0.03 / (L + 0.1) : L + 0.03 / (L + 0.1)
  const newC = L > 0.4 && !Number.isNaN(H) ? C * 0.93 + 0 : C
  const hoverColor = new Color('oklch', [newL, newC, H])

  let closestColor = hoverColor
  let minDistance = Infinity

  scales.forEach((scale) => {
    for (const color of scale) {
      const distance = hoverColor.deltaEOK(color)
      if (distance < minDistance) {
        minDistance = distance
        closestColor = color
      }
    }
  })

  hoverColor.coords[1] = closestColor.coords[1]
  hoverColor.coords[2] = closestColor.coords[2]

  return hoverColor
}

function transposeProgressionStart(
  to: number,
  arr: number[],
  curve: [number, number, number, number],
) {
  return arr.map((n, i, arr) => {
    const lastIndex = arr.length - 1
    const diff = arr[0] - to
    const fn = BezierEasing(...curve)
    return n - diff * fn(1 - i / lastIndex)
  })
}
