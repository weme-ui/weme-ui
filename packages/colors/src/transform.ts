import type { ColorAppearance, ColorScalesArray } from './types'
import { getColors } from './colors'
import { generateColor } from './generator'
import { getRadixColor, isRadixColorName } from './utils'

interface TransformColorOptions {
  color: string
  appearance?: ColorAppearance
  isNeutral?: boolean
  background?: string
}

export function transformColor({
  color,
  appearance = 'light',
  isNeutral = false,
  background = '#fff',
}: TransformColorOptions): ColorScalesArray {
  if (isRadixColorName(color))
    return getRadixColor(appearance, color, true)

  const colors = getColors(appearance, isNeutral)

  return generateColor(color, colors, background)
}
