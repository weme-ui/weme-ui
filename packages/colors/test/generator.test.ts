import { describe, expect, it } from 'vitest'
import { getColors } from '../src/colors'
import { generateColor } from '../src/generator'

describe('generator', () => {
  it('should get color objects', () => {
    expect(generateColor('#05f', getColors('light'), '#fff')).toMatchSnapshot()
    expect(generateColor('gray', getColors('dark', true), '#111')).toMatchSnapshot()
  })
})
