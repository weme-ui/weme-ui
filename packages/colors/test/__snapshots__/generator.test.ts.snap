// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generator > should get color objects 1`] = `
[
  "oklch(99.41% 0.0023 262.6)",
  "oklch(98.27% 0.011 262.6)",
  "oklch(96.06% 0.0233 262.6)",
  "oklch(93.56% 0.0456 262.6)",
  "oklch(90.37% 0.0686 262.6)",
  "oklch(86.31% 0.0944 262.6)",
  "oklch(80.72% 0.1159 262.6)",
  "oklch(73.09% 0.1525 262.6)",
  "oklch(53.32% 0.2596 262.6)",
  "oklch(48.58% 0.2366 262.6)",
  "oklch(51.8% 0.2366 262.6)",
  "oklch(31.46% 0.1181 262.6)",
]
`;

exports[`generator > should get color objects 2`] = `
[
  "oklch(17.76% 0 none)",
  "oklch(21.34% 0 none)",
  "oklch(25.38% 0 none)",
  "oklch(28.33% 0 none)",
  "oklch(31.3% 0 none)",
  "oklch(34.91% 0 none)",
  "oklch(40.22% 0 none)",
  "oklch(48.8% 0 none)",
  "oklch(59.99% 0 none)",
  "oklch(55.7% 0 none)",
  "oklch(77% 0 none)",
  "oklch(94.89% 0 none)",
]
`;
